torch/_inductor/codegen/cpp.py

```diff
@@ -1468,12 +1456,75 @@ def __init__(self, args, num_threads):
         self.reduction_depth = None
         self.reduction_prefix = IndentedBuffer()
         self.reduction_suffix = IndentedBuffer()
-        self.reduction_var_map = {}
+        self.parallel_reduction_prefix = IndentedBuffer()
+        self.parallel_reduction_suffix = IndentedBuffer()
+        self.parallel_reduction_stores: Dict[str, str] = dict()
+        self.local_reduction_init = IndentedBuffer()
+        self.local_reduction_stores = IndentedBuffer()
+        self.is_reduction = False
         self.reduction_cse = CSE(self.newvar_prefix, self.suffix, name_prefix="tmp_acc")
         self.preloads = IndentedBuffer()
         self.poststores = IndentedBuffer()
         self.num_threads = num_threads  # num_threads the kernel specialized for
         self.reduction_omp_dec: Dict[Tuple[str, str], str] = {}
+        self.max_threads = None
+
+    def _gen_parallel_reduction_buffers(
+        self,
+        acc,
+        acc_type,
+        reduction_type,
+        dtype,
+        value,
+        gen_store=True,
+        is_vec=False,
+    ):
+        if not self.max_threads:
+            self.parallel_reduction_prefix.writeline(
+                "int max_threads = omp_get_max_threads();"
```

Comment:
`parallel_reduction_prefix` appears before each worksharing scope. Would the `max_threads` be redefined when there are multiple worksharing scopes in a KernelGroup?

Reply:
Hi, Jiong, I am not clear about the question here.
From my understanding, the structure is:
```
parallel_reduction_prefix
worksharing1
{}
worksharing2
{}
```
Multiple worksharing will share 1 `max_threads` from 1 `parallel_reduction_prefix`. It should be correct?

Reply:
In your code, each parallel reduction kernel would define "max_threads" and then there might be multiple definitions?

Reply:
I failed to construct such a case.
But for code here https://github.com/pytorch/pytorch/blob/c7af626a2694eab6bce95a099d4c416e4aa90fb0/torch/_inductor/codegen/cpp.py#L1662-L1667
It looks only 1 kernels'  `prefix/suffix` will be returned as soon as `prefix/suffix` is found during the traverse within kernels.

Reply:
Have you tried to construct multiple kernels with parallel reduction in each of them. These kernels would be grouped into a single `KernelGroup` and might have redefined `max_threads` variable.

Reply:
I tryied
```
    @config.patch({"cpp.dynamic_threads": True})
    def test_reduction_with_dynamic_threads(self):
        def fn(a, b):
            return a.sum(), b.sum()

        self.common(
            fn,
            (torch.randn(1000), torch.rand(1000)),
        )
```
This will have 2 kernels in 1 KernelGroup. `max threads` are not redefined because there is
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py#L1717
So the `max thread` will be a local variable for 1 kernels code.
The code will be 
```
{
    {
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            int max_threads = omp_get_max_threads();
            float tmp_acc0_arr[max_threads];
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_arr[tid] = 0;}
            at::vec::Vectorized<float> tmp_acc0_vec_arr[max_threads];
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_vec_arr[tid] = at::vec::Vectorized<float>(0);}
            #pragma omp parallel
            {
                int tid = omp_get_thread_num();
                float tmp_acc0_local = 0;
                at::vec::Vectorized<float> tmp_acc0_vec_local = at::vec::Vectorized<float>(0);
                #pragma omp for
                for(long x0=static_cast<long>(0L); x0<static_cast<long>(992L); x0+=static_cast<long>(16L))
                {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(x0));
                    tmp_acc0_vec_local = tmp_acc0_vec_local + tmp0;
                }
                #pragma omp for
                for(long x0=static_cast<long>(992L); x0<static_cast<long>(1000L); x0+=static_cast<long>(1L))
                {
                    auto tmp0 = in_ptr0[static_cast<long>(x0)];
                    tmp_acc0_local = tmp_acc0_local + tmp0;
                }
                tmp_acc0_arr[tid] = tmp_acc0_local;
                tmp_acc0_vec_arr[tid] = tmp_acc0_vec_local;
            }
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0 = tmp_acc0 + tmp_acc0_arr[tid];
            }
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_vec = tmp_acc0_vec + tmp_acc0_vec_arr[tid];
            }
            tmp_acc0 = tmp_acc0 + at::vec::vec_reduce_all<float>([](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>& y) { return x + y; }, tmp_acc0_vec);
            out_ptr0[static_cast<long>(0L)] = static_cast<float>(tmp_acc0);
        }
    }
    {
        {
            float tmp_acc0 = 0;
            at::vec::Vectorized<float> tmp_acc0_vec = at::vec::Vectorized<float>(0);
            int max_threads = omp_get_max_threads();
            float tmp_acc0_arr[max_threads];
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_arr[tid] = 0;}
            at::vec::Vectorized<float> tmp_acc0_vec_arr[max_threads];
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_vec_arr[tid] = at::vec::Vectorized<float>(0);}
            #pragma omp parallel
            {
                int tid = omp_get_thread_num();
                float tmp_acc0_local = 0;
                at::vec::Vectorized<float> tmp_acc0_vec_local = at::vec::Vectorized<float>(0);
                #pragma omp for
                for(long x0=static_cast<long>(0L); x0<static_cast<long>(992L); x0+=static_cast<long>(16L))
                {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr1 + static_cast<long>(x0));
                    tmp_acc0_vec_local = tmp_acc0_vec_local + tmp0;
                }
                #pragma omp for
                for(long x0=static_cast<long>(992L); x0<static_cast<long>(1000L); x0+=static_cast<long>(1L))
                {
                    auto tmp0 = in_ptr1[static_cast<long>(x0)];
                    tmp_acc0_local = tmp_acc0_local + tmp0;
                }
                tmp_acc0_arr[tid] = tmp_acc0_local;
                tmp_acc0_vec_arr[tid] = tmp_acc0_vec_local;
            }
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0 = tmp_acc0 + tmp_acc0_arr[tid];
            }
            for (int tid = 0; tid < max_threads; tid++)
            {
                tmp_acc0_vec = tmp_acc0_vec + tmp_acc0_vec_arr[tid];
            }
            tmp_acc0 = tmp_acc0 + at::vec::vec_reduce_all<float>([](at::vec::Vectorized<float>& x, at::vec::Vectorized<float>& y) { return x + y; }, tmp_acc0_vec);
            out_ptr1[static_cast<long>(0L)] = static_cast<float>(tmp_acc0);
        }
    }
}
```

Reply:
Sounds good. BTW, there seems to be formatting issue in the code above. Please check:
```c++
for (int tid = 0; tid < max_threads; tid++) {
 tmp_acc0_arr[tid] = 0;}
```

Reply:
Fixed